cpu/rpx0xx/pio/i2c/i2c.c

```diff
@@ -0,0 +1,430 @@
+/*
+ * Copyright (C) 2021 Otto-von-Guericke Universität Magdeburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_rpx0xx
+ * @{
+ *
+ * @file
+ * @brief       PIO program to emulate an I2C interface
+ *
+ * @author      Fabian Hüßler <fabian.huessler@ovgu.de>
+ *
+ * @}
+ */
+#include <errno.h>
+#include "byteorder.h"
+#include "periph_conf.h"
+#include "periph/gpio.h"
+#include "periph/pio/i2c.h"
+#include "pio/pio.h"
+#include "i2c_set_scl_sda.hex.pio.h"
+#include "i2c.hex.pio.h"
+
+#define I2C_PIO_IRQN        0 /* use irq flag 0 */
+#define I2C_PIO_IRQN_SM(sm) ((1u << I2C_PIO_IRQN) << (sm))
+
+#define INSTR_I2C_SCL0_SDA0 (((pio_instr_t[])SET_SCL_SDA_PIO_PROGRAM)[0])
+#define INSTR_I2C_SCL0_SDA1 (((pio_instr_t[])SET_SCL_SDA_PIO_PROGRAM)[1])
+#define INSTR_I2C_SCL1_SDA0 (((pio_instr_t[])SET_SCL_SDA_PIO_PROGRAM)[2])
+#define INSTR_I2C_SCL1_SDA1 (((pio_instr_t[])SET_SCL_SDA_PIO_PROGRAM)[3])
+
+#define I2C_INSTR_FRAME(icount) ((uint32_t)(((uint32_t)(icount)) << 10))
+#define I2C_INSTR(instr) ((uint32_t)(instr))
+#define I2C_DATA_FRAME(final, data, nak)        \
+    ((uint32_t)((((uint32_t)(final)) << 9) |    \
+                (((uint32_t)(data)) << 1)  |    \
+                 ((uint32_t)(nak))))
+
+enum {
+    I2C_REPSTART = 0x40 /* make sure this flag is not used by the I2C driver */
+};
+
+static volatile uint32_t _irq[PIO_NUMOF];
+
+static void pio_i2c_init_pins(pio_t pio, pio_sm_t sm, gpio_t sda, gpio_t scl)
+{
+    const gpio_pad_ctrl_t pad_ctrl = {
+        .pull_up_enable = 1,
+        .input_enable = 1,
+        .drive_strength = DRIVE_STRENGTH_4MA,
+        .schmitt_trig_enable = 1,
+    };
+    const gpio_io_ctrl_t io_ctrl = {
+        .function_select = pio ? FUNCTION_SELECT_PIO1 : FUNCTION_SELECT_PIO0,
+        .output_enable_override = OUTPUT_ENABLE_OVERRIDE_INVERT
+    };
+
+    gpio_set_pad_config(scl, pad_ctrl);
+    gpio_set_pad_config(sda, pad_ctrl);
+
+    pio_sm_exec_block(pio, sm, pio_inst_set(PIO_INST_SET_DST_PINS, 0x3));
```

Comment:
this here needs some text to be understand able, thus why first writenig 3 then 0 again....

Reply:
It also works without `set pins 3` and later `set pins 0`. But the pico SDK does it like that too.
I guess that is what they mean with "avoid glitching". I now have copied their comment.

Reply:
my problem has, i did not knew that I have to setup my pio pins as output manually (to just write to them with the "out" instruction), thus after hours in the end I ended up looking into mp and pico sdk what they do - thus some comment here would have helped me a lot. i guess the pio interface should have something like "init pin" as output.
When I have my code in a clean shape, I can drop it as an example, and maybe i also get a clue howto do a init pin function, or even getting this compile pio step bit cleaner... at the moment I believe its can be made more convenient.
I used pio in mp before and i was that clean and easy, some of that must be possible here too ;-)